### ISSUE TITLE
Authoritative server 3.4.7 caches wrong SOA information in certain circumstances, leading to broken responses

### ISSUE THREAD

**Issue by seveas** [PowerDNS/pdns#3175]
b998fc0 and/or ef80925, possibly combined with existing behaviour, cause rather broken caching behaviour, which in turn causes subsequent queries to be answered with an empty answer until the cached entry expires.

The symptom:
```
spirit:~/pdns-bug$ dig +noall +answer +comments @localhost www1.c.example.com
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 60394
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
;; ANSWER SECTION:
www1.c.example.com.     600     IN      A       127.0.0.1

spirit:~/pdns-bug$ dig +noall +answer +comments @localhost DS www1.c.example.com
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 62140
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680

spirit:~/pdns-bug$ dig +noall +answer +comments @localhost www1.c.example.com
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 12800
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 2
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
```

As you can see, a DS query for a name causes subsequent queries for that name to get answers without AA bit and with 0 records. Even worse, the entire zone seems poisoned by the single DS query:

```
spirit:~/pdns-bug$ dig +noall +answer +comments @localhost
www1.c.example.com
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status:
NOERROR, id: 60394
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY:
0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
;; ANSWER SECTION:
www1.c.example.com.     600     IN      A       127.0.0.1

spirit:~/pdns-bug$ dig +noall +answer +comments @localhost DS www2.c.example.com
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 62140
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680

spirit:~/pdns-bug$ dig +noall +answer +comments @localhost www1.c.example.com
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 12800
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 2
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
```

In the logs of the backend I can see pdns querying the pipe backend for an SOA record of www1.c.example.com when pdns receives the DS query. The backend doesn't have such a record, so it prints END immediately. pdns then seems to cache this as the nonexistent SOA record for
the c.example.com zone, refusing to answer queries for any names in that zone until the broken cache entry expires. This is definitely new behaviour in 3.4.7 and I cannot reproduce it with 3.4.6. Whether that is because b998fc0 suddenly caches these or ef80925 has a bug, I don't know.

It does take a fairly specific setup to reproduce this, but this is reproducable reliably:

* PowerDNS 3.4.7, I've reproduced the issue with a fresh clone of the powerdns git tree and checking out tag 3.4.7, but also with RPM packages.
* negcache-ttl unset (so defaulting to 60 seconds)
* Pipe and bind backend in use. Pipe backend for a subdomain of a domain in the bind backend. I'm not sure whether it's relevant to use different backends, but I cannot reproduce the error if I don't enable the bind backend and stick a zonefile for example.com in there.

On https://gist.github.com/seveas/0101269a51ac765b887f you can find the exact pdns config, pdns bind backend config, zonefile and pipe backend implementation used to trigger this bug, they are minimized versions of the setup we have in production. After fixing up paths in the config that point into my homedir, you should be able to reproduce the bug with these.

I've tried to reproduce this issue with current master, but it seems that master is even more broken and/or requires the backend to change. When doing an ANY query for www1.c.example.com, the backend gets an SOA query for www1.c.example.com, and when the response of that is empty (which it's supposed to be), pdns doesn't ask the backend for an ANY or A response.

**Comment by pieterlexis** [comment]
Confirmed that this is an issue with the negative cache, as setting `negquery-cache-ttl=0` with these configs does not show this issue. As the code was ported to master, this issue also affects 4.0.0-alpha1

**Comment by ahupowerdns** [comment]
thanks for this detailed report & reproduction setup - very helpful. We are on it!

**Comment by seveas** [comment]
@pieterlexis that is correct, sorry for not mentioning that explicitly. We do not consider that an adequate workaround though, since we want to keep the negquery-cache. We downgraded to 3.4.6 for now.

**Comment by pieterlexis** [comment]
@seveas I'm not saying it a good work around, just so we know where to look exactly :smile:

**Comment by seveas** [comment]
@pieterlexis beware of red herrings though. The negquery-cache[-ttl] is actually doing its job. To me it looks more like the nonexistence of the SOA record for www1.c.example.com is stored under c.example.com.

**Comment by seveas** [comment]
Reverting ef80925 (improve SOA caching for DS queries) makes the issue disappear for me.

**Comment by seveas** [comment]
Ping. Any updates?

**Comment by pieterlexis** [comment]
Closed in #3770
